Move service-dependent methods out of database entity classes

ybnd opened the pull request:
## References
- Fixes https://github.com/DSpace/DSpace/issues/9027

## Description
This PR removes all database entity methods that used the corresponding Service, in order to improve separation of concerns & clarity.
- Many of the cases were remnants of the service refactor, but some others were added later
- Most of the methods in question were dealing with DSpaceObject metadata
- If similar methods are added in the future, they should be added to the service instead of the entity itself
- In general: if you find that you need to inject a service in a new entity method, you should move the whole method to the service instead

One of the most impactful changes is the removal of the `DSpaceObject.getName` method
- Since most DSO implementations depend on metadata for their name, this method cannot be part of the general interface for DSO entities
- Instead, the corresponding `DSpaceObjectService.getName` should be used
- The only exception to this is `Group`, which stores its name as a dedicated column

## Instructions for Reviewers
Confirm that the refactor is sensible and that everything works as it did before.
I've added some comments to highlight test fixes that may indicate more fundamental issues that we may want to investigate in more detail.

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome).
However, reviewers may request that you complete any actions in this list if you have not done so. If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is **created against the `main` branch** of code (unless it is a backport or is fixing an issue specific to an older branch).
- [ ] My PR is **small in size** (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
    - The refactoring performed in this PR touches a significant portion of the codebase
- [x] My PR **passes Checkstyle** validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR **includes Javadoc** for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR **passes all tests and includes new/updated Unit or Integration Tests** based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] My PR **includes details on how to test it**. I've provided clear instructions to reviewers on how to successfully test this fix or feature.
- [x] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [x] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

ybnd committed:
Move service-dependent DSpaceObject methods to appropriate services

+ deprecate DSpaceObject.getName in favour of DSpaceObjectService.getName
This will give us a lot more flexibility wrt. naming objects (e.g. to take into account the Entity type), but will require a lot of refactoring later.

ybnd committed:
Get rid of metadata methods on EPerson

ybnd committed:
Resolve compilation errors & test failures

ybnd committed:
Don't assume internal Item metadata cache is correct

This used to just work before, but seems unintentional

ybnd committed:
Simplify EPersonTest

- This test uses multiple services, so we can't consider it a unit test
- Removed boilerplate in favour of builders
- Removed old test cases that had been commented out for 12 years

ybnd committed:
Workaround for EPerson matcher failures

Seems like this used to happen before too; some EPersonRestRepositoryIT use matchEPersonOnEmail in the same way

ybnd committed:
Remove stray imports

ybnd committed:
Prepare automatic refactor for DSpaceObject.getName implementations

...by replacing all method bodies by direct calls to the specific DSO service's getName call.
The idea is to use IntelliJ's to inline these methods using IntelliJ, pulling the service call into the calling code like so:

  dso.getName() -> dso.dsoService.getName(dso)

Afterwards, this pattern can be targeted with a regex search-replace:

  ([^ ]+)\.(site|community|collection|...)Service.getName(\1) -> $2Service.getName($1)

 Some manual cleanup will also be needed to get everything to compile again, because:
 - The regex pattern above won't work if there are any line breaks in between
 - Many of the classes using these methods will have to updated to actually inject the services we need now

ybnd committed:
Inline *.getName to replace it with *Service.getName

ybnd committed:
Clean up after automatic refactor (until build works)

ybnd committed:
Further cleanup, fix tests

ybnd committed:
Use exact decimal format in QAEventMatcher

ybnd committed:
Remove leftover BitstreamFormat methods

ybnd commented:
As far as I can tell the new IT failure that comes up is unrelated to this PR; I've opened a separate PR for it: https://github.com/DSpace/DSpace/pull/9811

kshepherd commented:
This refactor looks great. @ybnd feel free to request review from me once you are ready for it (note you will need to rebase and push to get the IT to pass now that we fixed those issues)

ybnd commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/SubscriptionRestRepositoryIT.java:
```diff
@@ -1070,7 +1075,7 @@ public void linkedEpersonOfSubscriptionAdminTest() throws Exception {
         getClient(tokenAdmin).perform(get("/api/core/subscriptions/" + subscription.getID() + "/eperson"))
                              .andExpect(status().isOk())
-                             .andExpect(jsonPath("$", is(EPersonMatcher.matchEPersonEntry(eperson))));
+                             .andExpect(jsonPath("$", is(EPersonMatcher.matchEPersonOnEmail(eperson.getEmail()))));
```
Full entity machers now break these tests

ybnd replied:
Matchers can't always retrieve services from factories, so a simpler matcher is used here as a workaround.
I'll look into the larger problem, but in this particular case we don't really gain much by using the full matcher anyway.

See https://github.com/DSpace/DSpace/pull/9735#discussion_r1766419940

ybnd committed:
Reinstate CollectionService injection for getWorkflowGroup calls

kshepherd commented on dspace-swordv2/src/main/java/org/dspace/sword2/SimpleZipContentIngester.java:
```diff
@@ -142,8 +142,8 @@
-                bs.setFormat(context, format);
-                bs.setName(context, entry.getName());
+                bitstreamService.setFormat(context, bs, format);
+                bitstreamService.setName(context, bs, entry.getName());
```
While SWORD2 zip ingestion can be vulnerable to zip slip during extraction of the file itself, these changes are just about setting the metadata on the bitstream, so we can dismiss this alert.

kshepherd commented on dspace-api/src/main/java/org/dspace/content/DuplicateDetectionServiceImpl.java:
```diff
@@ -206,7 +206,8 @@ public Optional<PotentialDuplicate> validateDuplicateResult(Context context, Ind
         // Get item metadata and if it's configured for mapping, copy it across to the potential duplicate object
-        List<MetadataValue> metadata = resultItem.getCachedMetadata();
+        List<MetadataValue> metadata
+            = itemService.getMetadata(resultItem, Item.ANY, Item.ANY, Item.ANY, Item.ANY, true);
```
I think this should be OK, as `getMetadata` does return cached metadata if and only if the cache is not invalid. The original implementation was imperfect, agreed.

Why does it cause a problem after the refactor?

kshepherd reviewed (changes requested):
Thanks @ybnd , lots of tidying up! I went through every single file as a code review and also tested practical functionality in my local instance, and could not find any major issues.

I've added a few minor questions about a TODO item, some test inheritence, clarifying your comments. Once these comment threads are answered or resolved, I'm a +1 !

(I have no blocking changes requested, but I'm leaving the review in "Request changes" status just to reflect some open minor comments)

ybnd commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/matcher/WorkflowDefinitionMatcher.java:
```diff
@@ -26,7 +26,7 @@
  */
 public class WorkflowDefinitionMatcher {
-
+    // todo: this may not work in all cases!
```
Services don't always work nicely within matchers; IIRC this is also the reason why `SubscriptionRestRepositoryIT` has to match on email only.

I'll refresh _why_ this was the case exactly and at least document the distinction.

ybnd commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/matcher/QAEventMatcher.java:
```diff
@@ -80,7 +81,7 @@ public static Matcher<? super Object> matchQAEventNotifyEntry(QAEvent event) {
                     hasJsonPath("$.originalId", is(event.getOriginalId())),
                     hasJsonPath("$.title", is(event.getTitle())),
-                    hasJsonPath("$.trust", is(new DecimalFormat("0.000").format(event.getTrust()))),
+                    hasJsonPath("$.trust", is(QAEventRest.TRUST_FMT.format(event.getTrust()))),
```
Indeed, I updated the matcher to explicitly use the same formatting as the REST layer -- in my locale it used to break tests over `0.000` vs `0,000`

ybnd commented on dspace-api/src/test/java/org/dspace/eperson/EPersonTest.java:
I'll check out if `AbstractDSpaceObjectTest` is sufficient to start.

Honestly, I don't see a clear distinction between DSO and DSO service tests, and none of them are pure unit tests regardless of this PR -- IMO this is out of scope.

Commit:
Merge remote-tracking branch 'origin/main' into no-services-in-db-entities-9.0

Commit:
127675: Replace removed entity calls with service calls after latest main sync

Commit:
127675: Replace removed entity calls with service calls after latest main sync

Commit:
[TEMP] use services in matchers as they seem to work locally

ybnd committed:
Merge remote-tracking branch 'origin/main' into no-services-in-db-entities-9.0

ybnd committed:
Confirm that services factories work fine in Matchers

I can no longer replicate the original problems:
- Missing metadata values when matching EPerson created outside of builders (could be related to https://github.com/DSpace/DSpace/pull/10193)
- NPE when retrieving services from factories in class body

ybnd committed:
Minor mock setup fixes

kshepherd commented on dspace-api/src/main/java/org/dspace/content/LicenseUtils.java:
```diff
@@ -133,7 +136,7 @@ public static void grantLicense(Context context, Item item,
         AuthorizeException {
         // Put together text to store
-        // String licenseText = "License granted by " + eperson.getFullName()
+        // String licenseText = "License granted by " + epersonService.getFullName(eperson)
```
maybe not in scope for this PR, but this comment looks ancient and obsolete, we can probably remove it entirely (going by git history)

kshepherd reviewed (commented):
I inspected all the changed files. Looking great overall, I made ~10 small inline comments about some if nesting, TODO comments, var naming, etc.

And just one question about `EPersonTest` which had a big refactor and dropped a lot of individual tests (maybe they are covered by DSO Service testing?)

Just to clarify that I understand the motivation 100% - these methods we are shifting are primarily the ones that need decoupling because they can be implemented differently by service implementations, unrelated to the data model of the actual DSO? (e.g. `getName()` resolving a metdata query and concatenating, or `getWorkflowStep()` having several ways of determining this)

There are some other places I was less clear on, like wrapping `item.getBundles(...)` behind `itemService.getBundles(...)` (which does ultimately call `item.getBundles()` but not `bundle.getBitstreams()`. I also wondered whether we planned to wrap all accessors/mutators in services this way one day... would that make dev a bit easier if we didn't have to ask "where is the getter for X" and just always used the associated service? Or does that introduce needless complexity?
Is the motivation related to lazy vs eager loading of linked Hibernate resources? (presumably not entirely as item bundles are lazy, bundle bitstreams are lazy, and so on)

Anyway looks great, close to +1 once the small conflict is resolved, and I am hoping all my comments/questions are easy to address or dismiss. Thanks @ybnd !
